lib/chef/resource/powershell_package_source.rb

```diff
@@ -62,105 +156,167 @@ class PowershellPackageSource < Chef::Resource
+      # All PSRepositories are PackageSources, and all PackageSources with Provider PowerShellGet are PSRepositories. They are 2 different views of the same object.
+
+      action :register, description: "Registers a PowerShell package source." do
+        package_details = get_package_source_details
+        output = package_details.result
+        if output == "PSRepository" || output == "PackageSource"
+          action_set
+        elsif new_resource.provider_name.downcase.strip == "powershellget"
+          converge_by("register source: #{new_resource.source_name}") do
+            register_cmd = build_ps_repository_command("Register", new_resource)
+            res = powershell_exec(register_cmd)
+            raise "Failed to register #{new_resource.source_name}: #{res.errors}" if res.error?
           end
         else
-          if package_source_exists?
-            converge_if_changed :url, :trusted, :provider_name do
-              update_cmd = build_package_source_command("Set", new_resource)
-              res = powershell_exec(update_cmd)
-              raise "Failed to update #{new_resource.source_name}: #{res.errors}" if res.error?
-            end
-          else
-            converge_by("register source: #{new_resource.source_name}") do
-              register_cmd = build_package_source_command("Register", new_resource)
-              res = powershell_exec(register_cmd)
-              raise "Failed to register #{new_resource.source_name}: #{res.errors}" if res.error?
-            end
+          converge_by("register source: #{new_resource.source_name}") do
+            register_cmd = build_package_source_command("Register", new_resource)
+            res = powershell_exec(register_cmd)
+            raise "Failed to register #{new_resource.source_name}: #{res.errors}" if res.error?
+          end
+        end
+      end
+
+      action :set, description: "Updates an existing PSRepository or Package Source" do
+        package_details = get_package_source_details
+        output = package_details.result
+        if output == "PSRepository"
+          converge_if_changed :source_location, :trusted, :publish_location, :script_source_location, :script_publish_location, :source_name do
+            set_cmd = build_ps_repository_command("Set", new_resource)
+            res = powershell_exec(set_cmd)
+            raise "Failed to Update #{new_resource.source_name}: #{res.errors}" if res.error?
+          end
+        elsif output == "PackageSource"
+          converge_if_changed :source_location, :trusted, :new_name, :provider_name do
+            set_cmd = build_package_source_command("Set", new_resource)
+            res = powershell_exec(set_cmd)
+            raise "Failed to Update #{new_resource.source_name}: #{res.errors}" if res.error?
           end
         end
       end
 
       action :unregister, description: "Unregisters the PowerShell package source." do
-        if package_source_exists?
```

Comment:
The :unregister semantics changed here so that it no longer checks if the package_source_exists and will update and attempt to unregister the package source every time.

Reply:
I brain-cramped when we talked about this during that review. There is a checker in there to make sure that a package source exists (or not) before undertaking a register or unregister action